### DISCUSSION TITLE
Scramjet engine design

### DISCUSSION THREAD

**Discussion by Jacki43** [JSBSim-Team/jsbsim#395] [Support requests]
Hello all,

I want to model a scramjet engine with c++ software, and I am a beginner with JSBSIm. I wanted to ask you is it possible in JSBSim to model a scramjet engine? 
Can you please give me a brief guide on how can I do that?
Thanks in advance

**Comment by seanmcleod** [comment]
I suggest you take a look at the source code for the existing engine models in `src\models\propulsion` as a first step to get familiar with how engine models are integrated in JSBSim. Which should give you an idea of how to go about adding a scramjet engine model.

**Comment by Jacki43** [comment]
Thank you for your reply.
I think the engine model that exists in JSBSim which is similar to the scramjet engine, is FGTurboProp.cpp. Because both of them are airbreathing. Consequently, I think I should eliminate the compressor and turbine that are used in the turboprop model and add the nozzle and diffuser into that to implement a scramjet engine. Do you agree with that?
Is there any idea on how to consider diffuser and nozzle to implement scramjet?

Thanks in advance

**Comment by seanmcleod** [comment]
I was suggesting more that you look at the engine models to get an idea of the classes involved and the methods you need to implement. I wouldn't try and start will all of the `FGTurboProp` or `FGTurbine` code and try and remove the compressor modeling.

What is the model (set of equations) you have for a scramjet, what are the inputs required including configuration options to then run through your equations to calculate the amount of thrust given the inputs?

**Comment by bcoconni** [comment]
> I think the engine model that exists in JSBSim which is similar to the scramjet engine, is FGTurboProp.cpp. Because both of them are airbreathing. 

Please don't do that. I would recommend against using `FGTurboProp.cpp` or `FGTurbine.cpp` as a start for your scramjet model.
> Consequently, I think I should eliminate the compressor and turbine that are used in the turboprop model and add the nozzle and diffuser into that to implement a scramjet engine. Do you agree with that?

You can't. There are no compressor/turbine/nozzle in `FGTurboProp.cpp` and `FGTurbine.cpp`. They are not engineering models. You will not find any physics/thermodynamics model in them.

As @seanmcleod said, your best course of action is to build a new model from the ground up using the set of equations you have.

Basically, you need to:
* Write a new class `Scramjet` (or any name that you see fit) that inherits from [`FGEngine`](https://github.com/JSBSim-Team/jsbsim/blob/master/src/models/propulsion/FGEngine.h)
* Declare the new type of engine in `FGPropulsion.cpp`. For that, you need to add a new `if` condition with your model name `scramjet` and call the constructor of your class as in the code extract below.
https://github.com/JSBSim-Team/jsbsim/blob/c68ef9f26f55ed9920f66bd12a9b2fb2ed40ef38/src/models/FGPropulsion.cpp#L417-L425
* For a start you can skip fuel management and use stub methods for anything that is linked to fuel and tanks.
* You should rather use the `FGElectric` class for an example as it contains the bare minimum for an `FGEngine` class derivation.
  * In `FGElectric` you will see that the method `Calculate` is the place where the engine parameters are computed with respect to the throttle command, atmospheric properties, etc.

**Comment by Jacki43** [comment]
Thank you so much @bcoconni, @seanmcleod.

I will consider your suggestion to design a scramjet engine. However, at first, I want to clarify something for myself. As you know, the `JSBSim-master\engine\PW125B.xml` is a turbojet. I wanted to know the trust equation of this [turbojet ](https://www.grc.nasa.gov/www/k-12/airplane/turbth.html)has been defined in which place of the code?
You mentioned that: 

> They are not engineering models. You will not find any physics/thermodynamics model in them

So, how the trust force of PW125B has been defined in JSBSim?

I wanted to apply the trust equation of [scramjet](https://www.grc.nasa.gov/www/k-12/airplane/ramth.html), but I do not know where can I define the equation and parameters (for instance, mass flow rate, pressure, velocity, and area for stream and exit conditions). I thought that if I find the source files of turbojet and try to change the trust equation of turbojet to the equation of scramjet, I will have an initial result for scramjet engine design.
Can you please let me know if there is an idea about that @bcoconni?

Thanks in advance

**Comment by seanmcleod** [comment]
As @bcoconni mentioned use `FGElectric` as an example/template for your `FGScramjet` class. Also take a look at the `FGTurbine` class to see how it makes use of tables/functions in case you need that sort of functionality.

Looking at the NASA example you linked to:

![image](https://user-images.githubusercontent.com/824856/107113402-6b891c00-6867-11eb-89a6-ca09c2fdf028.png)

At a minimum you would need to read in configuration parameters for the area of the exit nozzle Ae and area of the inlet nozzle etc. Here is an example showing how `FGTurbine` reads in the bypass ratio configuration option.

https://github.com/JSBSim-Team/jsbsim/blob/33d429e783589c27dccf2d70d231efe98c38f1df/src/models/propulsion/FGTurbine.cpp#L435-L436

Then in your `FGScramjet::Calculate` method you would make use of the configuration values plus make use of the input commands like the throttle command and atmospheric properties to calculate the thrust as per the equation shown in the NASA example.

The throttle command and atmospheric inputs are available via:

https://github.com/JSBSim-Team/jsbsim/blob/33d429e783589c27dccf2d70d231efe98c38f1df/src/models/propulsion/FGEngine.h#L117-L142

Then in `FGScramjet::Calculate` you'll have:

```c++
thrust = NASA equation;
Thruster->Calculate(thrust);
```

**Comment by Jacki43** [comment]
Thank you so much for your detailed response. 

> Then in your `FGScramjet::Calculate` method you would make use of the configuration values plus make use of the input commands like the throttle command and atmospheric properties to calculate the thrust as per the equation shown in the NASA example.

As you mentioned, I applied the codes that were related to the bypass ratio configuration to the FGSceamjet. However, when I want to use the atmospheric properties in the thrust equation, I have the error: the properties are undefined. Is there any suggestion on how to read these properties? Should be added specific library to the FGScramjet?

Thanks in advance

**Comment by bcoconni** [comment]
> However, when I want to use the atmospheric properties in the thrust equation, I have the error: the properties are undefined. Is there any suggestion on how to read these properties?

Could you please show us the lines of code that cause the compilation errors ?
Could you also show the error messages that you get ?

**Comment by Jacki43** [comment]
Thank you for your reply. I only applied some changes in the FGElectric file that you mentioned to implement FGScramjet.
The line of code that I got error  is `  Thruster->Calculate(vt-vc);` . Please note that `vt-vc` is a sample thrust formula to test that whether the atmospheric properties can be read.
`
void FGScramjet::Calculate(void)
{
  RunPreFunctions();


  if (Thruster->GetType() == FGThruster::ttPropeller) {
    ((FGPropeller*)Thruster)->SetAdvance(in.PropAdvance[EngineNumber]);
    ((FGPropeller*)Thruster)->SetFeather(in.PropFeather[EngineNumber]);
  } 

  RPM = Thruster->GetRPM() * Thruster->GetGearRatio();

  LoadThrusterInputs(); 
  Thruster->Calculate(vt-vc);

  RunPostFunctions();
`

Also, the error is:
<img width="399" alt="error" src="https://user-images.githubusercontent.com/78500616/107161888-a2e1f080-696d-11eb-9319-cc7e9bcc6c7b.png">

**Comment by seanmcleod** [comment]
`Vt` and `Vc` are not member variables of `FGEngine`, take a look at the code snippet I shared earlier, they're part of a `struct`. If you take a look at the constructor for `FGElectric` or any of the other engine models you'll see a reference is passed to this struct.

```c++
FGElectric::FGElectric(FGFDMExec* exec, Element *el, int engine_number, struct FGEngine::Inputs& input)
  : FGEngine(engine_number, input)
{
```

Which is then passed to the constructor of `FGEngine` which is the base class. If you take a look at the `FGEngine` class you'll see it has the following member variable.

```c++
  struct Inputs& in;
```

You'll see this member variable `in` actually being used in the propeller code you copied.

Why have you copied the propeller related code into your scramjet model?

**Comment by seanmcleod** [comment]
@Jacki43 I had mentioned in my first response - 

> I suggest you take a look at the source code for the existing engine models in src\models\propulsion as a first step to get familiar with how engine models are integrated in JSBSim. 

The idea being to understand how engine models are structured, what base classes to inherit from, where and how they get their configuration data, where and how they get their required input parameters from etc.

**Comment by Jacki43** [comment]
Thank you so much for your response. I got the point.
I should use the `Vt`, `Vc` in the formula as follows: 
`Thruster->Calculate(in.Vt-in.Vc);`
I tested it and I do not have any error right now.

> Why have you copied the propeller related code into your scramjet model?

I just wanted to try to apply the thrust equation, so I did not check all of the things. As you mentioned, I should go through the sample engine codes and make sure I am designing the scramjet engine in the right way. 

It will be great if this issue stays open so that I can ask my questions and get your opinions that are very useful for me.
Thanks in advance

**Reply by bcoconni** [reply]
> I should use the `Vt`, `Vc` in the formula as follows: 
`Thruster->Calculate(in.Vt-in.Vc);`

Not quite so. The `Thruster` instance must be supplied with a thrust not a velocity.
In addition, I am pretty sure the difference between the true airspeed `in.Vt` and the calibrated airspeed `in.Vc` is irrelevant to the computation of a ramjet thrust.

**Comment by Jacki43** [comment]
Thank you so much for your help.

I am trying to implement the scramjet NASA thrust equation as I mentioned. However, all of the variables that I need, do not exist in `jsbsim/src/models/propulsion/FGEngine.h`. I need the pressure, mass flow rate, and velocity of both stream and exit conditions. 
Do you think for the exit condition (nozzle), should I use the variables that exist in `jsbsim/src/models/propulsion/FGNozzle.cpp`? 
I think the velocity of stream condition is `Vt`. What about the velocity of exit condition, and what is its variable?

Thanks in advance

**Reply by bcoconni** [reply]
> I need the pressure, mass flow rate, and velocity of both stream and exit conditions.

You will have to [compute them](https://web.mit.edu/16.unified/www/FALL/thermodynamics/notes/node85.html#SECTION06363000000000000000). The mass flow rate being constrained by the nozzle exit area which I guess is always chocked for a ramjet.

**Comment by seanmcleod** [comment]
Yes at some point in most cases you won't have every single variable available but will need to calculate and derive the variables you need from the existing variables.

So for example how would you derive the incoming mass flow rate?

**Comment by Jacki43** [comment]
Thank you for your reply.
I rewrote the thrust equation in such a way that I use the existing variables. However, before applying I wanted to ask you please let me know what are the` Pressure ` and ` Density `. Are they related to the exit condition?
I found a [description](https://sourceforge.net/p/jsbsim/mailman/jsbsim-users/thread/CACdXh0C0_BQ9RifMmMskoF5mps5CkPRSZuyutk-LmYTVt0BQ0g@mail.gmail.com/) for `Vt` :

>  velocities/vt-fps is the total airspeed (in feet per second) : it is the
norm of the aircraft velocity with respect to the air measured in the body
frame.

Consequently, we have `Ve= Vt+V0`. Is it true?
Thanks in advance

**Reply by bcoconni** [reply]
> However, before applying I wanted to ask you please let me know what are the` Pressure ` and ` Density `. Are they related to the exit condition?

`Pressure` and `Density` are the air pressure and density just upstream of the aircraft. From the scramjet point of view, `Pressure` is the static pressure. You'll need to convert it to stagnation pressure to compute the conditions at the entry of the scramjet. And you will have to do the same for the temperature.

Obviously these conditions colud be significantly different at the exit. In the case of an ideal ramjet:
* The pressure will be the same at the exit than it was at the entry/inlet
* The temperature will be higher
* The density will be lower (as a result of the ideal gas relationship `P=rho*R*T`)
> Consequently, we have `Ve= Vt+V0`. Is it true?

True airspeed `Vt` and Equivalent airspeed `Ve` are not linked by a linear relationship as can be seen in [Wikipedia article on true airspeed](https://en.wikipedia.org/wiki/True_airspeed#Calculating_true_airspeed)

**Comment by seanmcleod** [comment]
> let me know what are the `Pressure` and `Density`. Are they related to the exit condition?

Did you search through the code to see where and how they're set?

https://github.com/JSBSim-Team/jsbsim/blob/193d28c20788bfd6d6daa27a9573c8a16d92acd8/src/FGFDMExec.cpp#L501-L525

So they're the free stream values, the 0 subscript values in the NASA equation.

Also, think about it, how would the generic base class `FGEngine` know anything about your specific engine to figure out how to calculate an exit value? What if you're modeling an electric engine?

**Comment by Jacki43** [comment]
Thank you for your quick response. 

> Also, think about it, how would the generic base class FGEngine know anything about your specific engine to figure out how to calculate an exit value? What if you're modeling an electric engine?

I did not find `jsbsim/src/FGFDMExec.cpp`, so I was confused. 
Thank you so much.

**Reply by bcoconni** [reply]
> I did not find `jsbsim/src/FGFDMExec.cpp`, so I was confused. 

May be you could search for `src/FGFDMExec.cpp`?

**Reply by Jacki43** [reply]
Thank you for your reply @bcoconni. I searched and found it. Just one question is that what is the property of "direct" trust value that I could put in notify section and get its value during the flight?
When I use `<property>propulsion/engine/thrust-lbs</property>`, its value is 0. 

Also, the value of the properties ` <property>atmosphere/sigma</property>` and `<property>atmosphere/rho-slugs_ft3</property>` that I used them in the thrust equation, are 0. I do not have any errors in my code. However, when I change the engine type from the rocket engine to the electric engine, the value of these properties becomes 0. Do you have any idea that this problem is related to which part of the code?

**Reply by bcoconni** [reply]
> Just one question is that what is the property of "direct" trust value that I could put in notify section and get its value during the flight?
When I use `<property>propulsion/engine/thrust-lbs</property>`, its value is 0. 

It is most likely because you did not pass the thrust value (in pound-force) to `Thruster` in your C++ class:
```c++
  Thruster->Calculate(thrust);
```
> Also, the value of the properties ` <property>atmosphere/sigma</property>` and `<property>atmosphere/rho-slugs_ft3</property>` that I used them in the thrust equation, are 0.

- Air density in slug/ft^3 is available from the variable `in.Density` in your C++ class.
- `sigma` a.k.a. density ratio is available from `in.DensityRatio`.

**Reply by Jacki43** [reply]
Thanks for your reply.

> It is most likely because you did not pass the thrust value (in pound-force) to Thruster in your C++ class.

I had defined the thruster equation in `Thruster` in my code as follows. Do you think is there anything wrong with this part of the cod?
![Capture1](https://user-images.githubusercontent.com/78500616/108584642-1215f800-7311-11eb-937c-69e3c586c165.JPG)

**Reply by bcoconni** [reply]
I see a number of errors in your code:
- `in.PressureRatio` is the ratio of the ambient pressure to the sea level pressure.
https://github.com/JSBSim-Team/jsbsim/blob/d528f8b38effb54ee035b572ee1beebdea7d2eb8/src/models/FGAtmosphere.h#L152-L153
- `in.DensityRatio` is the ratio of the ambient density to the sea level density.
https://github.com/JSBSim-Team/jsbsim/blob/d528f8b38effb54ee035b572ee1beebdea7d2eb8/src/models/FGAtmosphere.h#L175-L176

- I guess that `pe`, `re`, `ve`, `Ae` and `mdote` are respectively the pressure, density, velocity, area and mass flow at the ramjet exit ? If that is indeed the case then your equations are incorrect. Since you are modelling a perfect ramjet, it is assumed that there are no pressure losses and that there are no compressors/turbines therefore **the stagnation pressure at exit equals the stagnation pressure at inlet**. In addition, as explained in the [MIT site that I mentioned above](https://web.mit.edu/16.unified/www/FALL/thermodynamics/notes/node85.html#SECTION06363000000000000000), the **inlet and exit Mach numbers are also the same**. Hence the inlet and exit static pressure are equal as well i.e. `pe == po` at any time.

- The exit temperature `Te` is a result of the temperature rise that occurred in the combustion chamber. I see no such computation of the temperature in your model.

- The exit density `re` is then obtained from the [ideal gas law](https://en.wikipedia.org/wiki/Ideal_gas): `re=pe/(R*Te)`

- The computation of the exit velocity `ve` is using the [Bernoulli's principle for incompressible fluids](https://en.wikipedia.org/wiki/Bernoulli%27s_principle#Incompressible_flow_equation) which is incorrect. Gases are generally assumed to be incompressible for Mach numbers lower than 0.3 but a ramjet is a propulsion device for supersonic/hypersonic vehicles so the hypothesis of incompressibility falls.

- The computation of the exit area `Ae` breaks the conversation of mass. The form that you are using could be valid for incompressible fluids (i.e. when `re == ro`) but, as explained above, this assumption is not met. So you should have `re*Ae*ve == ro*Ao*vo` at any time.

Have you checked that the value of `Thrust` is not zero ? If you are running your simulation at sea level then `in.DensityRatio` and `in.PressureRatio` are equal to `1.0` which should explain why `Thrust` is equal to `0.0`.

**Comment by bcoconni** [comment]
This issue (#392) has been moved to `Discussions`.

Support requests and questions are now resolved in [`Discussions`](https://github.com/JSBSim-Team/jsbsim/discussions) while [`Issues`](https://github.com/JSBSim-Team/jsbsim/issues) are restricted to bugs reports and feature requests.

This change is only motivated by project organization and will not affect the support provided. We'll gladly continue to answer your questions 😄.

**Comment by seanmcleod** [comment]
I was also going to mention that a number of the parameter calculations didn't seem to make sense, but my main question is what sort of debugging have you tried?

In particular if you set a breakpoint on line 120 where the thrust is calculated what value do you see calculated? If it is 0 then that answers your question below. If it is 0 then go back and check the parameters you use to calculate thrust, do they make sense, are there calculation bugs?

> When I use <property>propulsion/engine/thrust-lbs</property>, its value is 0.

If your thrust calculation isn't 0 then step into `Thruster->Calculate(thrust)`. What do you see happening in that method in the debugger?

What thruster do you have connected in your model? I'm a bit suspicious that you've added your code to `FGElectric` and so I'm wondering what sort of thruster you have connected. Why haven't you implemented your own `FGRamJet` class?

**Comment by Jacki43** [comment]
Thank you so much for your detailed answer @bcoconni. My assumption was wrong and you are right. I applied some new equations with the help of [this website](https://ww2.mathworks.cn/help/aerotbx/ug/determining-heat-transfer-and-mass-flow-rate-in-a-ramjet-combustion-chamber.html) and [this one](https://www.grc.nasa.gov/www/k-12/VirtualAero/BottleRocket/airplane/nozzleh.html), so the value of thrust has not been zero. However, when I apply this scramjet model to X15 (instead of the XLR99 that it is already used), the value of thrust in the "climb" event is: ` propulsion/engine/thrust-lbs = -265.837163`. It is very low and negative. I wanted to make sure that whether this result is related to the equations that I applied or any other part of the code. Consequently, please let me describe here the changes that I applied to have a scramjet engine @seanmcleod . 

I changed the engine and thruster type as follows:
```xml
    <propulsion>
        <engine file="scramjet">
            <location unit="IN">
                <x> 600 </x>
                <y> 0 </y>
                <z> 0 </z>
            </location>
            <orient unit="DEG">
                <roll> 0.0 </roll>
                <pitch> 0 </pitch>
                <yaw> 0 </yaw>
            </orient>
            <feed>0</feed>
            <feed>1</feed>
            <thruster file="direct">
                <location unit="IN">
                    <x> 600 </x>
                    <y> 0 </y>
                    <z> 0 </z>
                </location>
                <orient unit="DEG">
                    <roll> 0.0 </roll>
                    <pitch> 0.0 </pitch>
                    <yaw> 0.0 </yaw>
                </orient>
            </thruster>
        </engine>
        <tank type="FUEL">    <!-- Tank number 0 for Main engine -->
            <location unit="IN">
                <x> 282.3 </x>
                <y> 0 </y>
                <z> 0 </z>
            </location>
            <capacity unit="LBS"> 9470 </capacity>
            <contents unit="LBS"> 0 </contents>
        </tank>
        <tank type="FUEL">    <!-- Tank number 1 for Main engine -->
            <location unit="IN">
                <x> 408.3 </x>
                <y> 0 </y>
                <z> 0 </z>
            </location>
            <capacity unit="LBS"> 8236 </capacity>
            <contents unit="LBS"> 6 </contents>
        </tank>
        <tank type="FUEL">    <!-- Tank number 2 for RCS -->
            <location unit="IN">
                <x> 140 </x>
                <y> 0 </y>
                <z> 0 </z>
            </location>
            <capacity unit="LBS"> 12 </capacity>
            <contents unit="LBS"> 0 </contents>
        </tank>
    </propulsion>
```
The content of scramjet.xml engine file is:

```xml
      <electric_engine name="scramjet">
         <augmented>         0  </augmented>
         <injected>          0  </injected>
       </electric_engine>
```
Finally, the equations that I applied for thrust are:

![Capture2](https://user-images.githubusercontent.com/78500616/108793218-087ad300-7551-11eb-9cae-8a115eb8d6e9.JPG)

**Reply by bcoconni** [reply]
Have you checked your units ?
* The gas constant `R` is `287.1` in `J/K/kg` (I'm surprised to read `2871` which seems 10 times as much ?)
* `in.Pressure` is given in `psf` (pound-force per square feet)
* `in.Temperature` is given in Rankine
* `in.Vt` in feet per second
* `cp` in SI units is `gama*R/(gama-1)` which is approximately 1004 in `J/K/kg`

So `ro` is computed from quantities expressed in a mixture of SI units and imperial/US/UK units and is most likely incorrect. In any case the air density is already available from the variable `in.Density` which is expressed in `slug/ft^3`. And since the variable `mdot` in your program is obtained from `ro` then the mass rate is most likely incorrect as well.

I would rather use a constant value for `te` because this is how a combustion chamber basically works: at full thrust you burn as much fuel as you can to produce the maximum heat that the materials can withstand. This is especially true for a scramjet where you are not limited by the maximum rotational speed that the rotors can withstand.

**Reply by bcoconni** [reply]
According to the site you mentioned `NPR` should be higher than 1.0 (total pressure is **always** higher than static pressure) and can easily be computed (assuming there are no pressure losses in the scramjet): `NPR = in.TotalPressure/in.Pressure`.

Also I would leave efficiencies apart for a start (`eta = 1.0`).

**Reply by bcoconni** [reply]
I guess you replaced `1-(1/NPR)^(gamma-1)/gamma` by 0.03 because you got a negative number using `NPR=0.9` ?

**Comment by seanmcleod** [comment]
> I wanted to make sure that whether this result is related to the equations that I applied

So when you set a breakpoint on `Thrust = mdot * (ve - vo);`  what do you see? If thrust is negative then either `mdot` is negative or `ve - vo` is negative.

**Reply by Jacki43** [reply]
Thank you for your reply.
When I set a breakpoint on the thrust equation line, I have the following results; however, the value of thrust (-9.2559631349317831e+61) is not equal to `mdot * (ve - vo) = 0.265 * (10.88-756.63) = -197.62`. Also, In the X153.xml, in `simulation/notify-time-trigger" value="0.1"`, I have `propulsion/engine/thrust-lbs = -197.6`.
I am totally confused what are those thrust values that I got!

![Capture4](https://user-images.githubusercontent.com/78500616/108874364-17519c00-75ca-11eb-8927-e65e6b705f53.JPG)

**Reply by bcoconni** [reply]
Obviously as @seanmcleod said, `ve` is lower than `vo` which means you cannot produce thrust.

**Reply by Jacki43** [reply]
Thank you so much for your reply @bcoconni and @seanmcleod. Yes, I checked the units of parameters, and I used [these](https://www.studocu.com/en-us/document/georgia-institute-of-technology/jet-rocket-propulsion/other/hw4soln-homework-4-solution/5780265/view) equations for thrust, so the problem has been solved with your help.

![Capture5](https://user-images.githubusercontent.com/78500616/109364567-12584b00-785d-11eb-9367-70624c1f4f29.JPG)


![Capture6](https://user-images.githubusercontent.com/78500616/109364576-15ebd200-785d-11eb-8889-339d33de7cad.JPG)

Thank you so much.

**Reply by Jacki43** [reply]
Hello all,
I faced two issues.

-First, when I use the XLR99 engine, the value of Mach is 4 (in Apogee) while in the scramjet engine, Mach is less than 1. The scramjet should produce more thrust than the rocket engine, so the velocity should be more than that! I do not know is from which part of my code. When I multiply thrust by 10, the velocity reaches Mach 4.

-Furthermore, the `propulsion/tank/contents-lbs` do not decrease during the flight and are constant. Are there specific properties for this task that should be defined in the engine script file?

Is there any idea on these issues? 
The following codes show the comparison of x15 in two different runs (one with XLR99 engine and another one with scramjet engine). 

**Apogee event ( run with XLR99 engine):**

> Apogee (Event 5) executed at time: 187.430000
    velocities/mach = 4.141563
    velocities/h-dot-fps = -0.198912
    position/h-agl-ft = 274216.626387
    attitude/pitch-rad = -0.013265
    fcs/throttle-pos-norm = 0.000000
    propulsion/engine/thrust-lbs = 0.000000
    propulsion/tank[0]/contents-lbs = 0.000000
    propulsion/tank[1]/contents-lbs = 59.514250
    propulsion/tank[2]/contents-lbs = 12.000000

**Apogee event ( run with scramjet engine):**

> Apogee (Event 5) executed at time: 71.220000
    velocities/mach = 0.725899
    velocities/h-dot-fps = -0.060488
    position/h-agl-ft = 48040.043519
    attitude/pitch-rad = 0.278655
    fcs/throttle-pos-norm = 1.000000
    propulsion/engine/thrust-lbs = 7550.042961
    propulsion/tank[0]/contents-lbs = 9400.000000
    propulsion/tank[1]/contents-lbs = 8236.000000
    propulsion/tank[2]/contents-lbs = 12.000000

Thanks in advance.

**Reply by seanmcleod** [reply]
Well there are massive differences in this single data point (apogee) right?

The XLR99 combination took 187s to reach apogee, and it reached 274,261ft in altitude, versus only 71s and an altitude of only 48,040ft for your scramjet combination.

So given these massive differences I wouldn't be concerned about the difference in Mach at apogee.

Are you logging the data at even just 1Hz and comparing the thrust produced for the two combinations to compare? My guess (given the very limited data you've provided) is that your scramjet is producing a lot less thrust all the time compared to the XLR99.

**Reply by bcoconni** [reply]
> First, when I use the XLR99 engine, the value of Mach is 4 (in Apogee) while in the scramjet engine, Mach is less than 1. The scramjet should produce more thrust than the rocket engine, so the velocity should be more than that! I do not know is from which part of my code. When I multiply thrust by 10, the velocity reaches Mach 4.

Ramjets are notoriously inefficient at low Mach numbers / subsonic velocities. You can plot your ramjet thrust against Mach to convince yourself: you should see a bell curve with a peak in the vicinity of Mach 2 or 3 and thrust decreasing when flying above or below that velocity like below.

![fig5RamjetPerformance_web](https://user-images.githubusercontent.com/7502904/111028927-2a51d200-83fa-11eb-9ccf-21edfe503f35.jpg)


This is the reason why vehicles equipped with ramjets are first accelerated with a rocket-like motor before the ramjet is ignited.

The other reason might be that your ramjet peak thrust is too low. In that case, you have 2 options to increase the thrust:
1. Increase the combustion temperature to a higher value (increase the value of `tmax` - according to the code you published earlier). This will increase the exhaust velocity and therefore increase the thrust.
2. Increase the mass flow by increasing the engine section `Ao`. Same as above: this will result in the increase of `mdot` and therefore increase the thrust as well.
> Furthermore, the `propulsion/tank/contents-lbs` do not decrease during the flight and are constant. Are there specific properties for this task that should be defined in the engine script file?

For that you need to implement the fuel consumption in the method `CalcFuelNeed` which must return the amount of fuel (in pounds) that has been burnt in the last time step `in.TotalDeltaT`
https://github.com/JSBSim-Team/jsbsim/blob/518f2a74d1a5db65de3986816d3e6cd90521585e/src/models/propulsion/FGTurbine.cpp#L380-L386

**Reply by Jacki43** [reply]
> Increase the mass flow by increasing the engine section Ao. Same as above: this will result in the increase of mdot and therefore increase the thrust as well.

Thank you so much for your detailed answer. I had this idea in my mind, but I was not sure whether it is reasonable. As you mentioned, I increased the `Ao` and I got the result that I wanted.

> For that you need to implement the fuel consumption in the method CalcFuelNeed which must return the amount of fuel (in pounds) that has been burnt in the last time step in.TotalDeltaT

As you mentioned, I considered the following code in my FGElectric.cpp file. However, I do not have fuel consumption yet. Is there any idea regarding this issue? 
![Capture6](https://user-images.githubusercontent.com/78500616/112064388-40b60700-8b39-11eb-8d6e-678bfe7b11d3.JPG)

**Reply by bcoconni** [reply]
Basically fuel is used to heat up the air in the combustion chamber so you need a bit of [thermodynamics](https://en.wikipedia.org/wiki/Thermodynamics) to evaluate how much fuel is used.
1. The amount of energy stored in the air is the [enthalpy](https://en.wikipedia.org/wiki/Enthalpy) which is basically `H=mdot*Cp*T` where `T` is the gas total temperature, `Cp` the specific heat at constant pressure and `mdot` the air flow.
2. The combustion chamber increases the enthalpy by burning a fuel quantity `mdot_fuel` (in kg/s) which has a [Lower Heating Value](https://en.wikipedia.org/wiki/Heat_of_combustion#Lower_heating_value) `Q_fuel` (in J/kg).
3. So the air enthalpy `H` increases by `mdot_fuel*Q_fuel` when being heated up in the combustion chamber.
4. Using your variable names, the specific enthalpy before the combustion chamber is `Cp*to` and after the combustion chamber is `Cp*te`. So we have `mdot_fuel*Q_fuel=mdot*(Cp*te-Cp*to)` and finally `mdot_fuel=Cp*(te-to)*mdot/Q_fuel`.

You have already evaluated `Cp` to be `1004` (J/kg/K) and `Q_fuel` is typically `43e6` (J/kg) for kerosene (search "Lower Heating Value" on internet for other fuels such as methane, hydrogen, etc.).

Note that `CalcFuelNeed` is expected the return a mass of fuel in `lbs` so you need to integrate the fuel flow and convert the result to pounds if needed:
```c++
double FGElectric::CalcFuelNeed(void)
{
    const double Q_fuel = 43e6; // Kerosene (J/kg)
    FuelFlowRate = mdot*Cp*(te-to)/Q_fuel; // kg/s

    FuelFlowRate *= kgtolb ; // Convert from kg/s to lb/s
    FuelExpended = FuelFlowRate * in.TotalDeltaT;
    if (!Starved) FuelUsedLbs += FuelExpended;
    return FuelExpended;
}
```
Note that this is a very simplified evaluation of what actually is going on in a combustion chamber but I think this should meet your needs for a start.

**Comment by seanmcleod** [comment]
When the breakpoint is hit, the line hasn't been executed yet. The debugger is stopping the program before it executes the line, so the value you're seeing for the local variable `Thrust` is some uninitialized value.

Execute the line, in Visual Studio hit F-10 which will execute the line and then stop on the next line. But as you can see, `ve < vo`.
